Add NCCL Alltoall to PT NCCL process group

srinivas212 opened the pull request:
Summary: Add Alltoall and Alltoallv to PT NCCL process group using NCCL Send/Recv.

Differential Revision: D20781624

mrshenli commented:
Hey @srinivas212, thanks for contributing! @agolynski is working on revamping c10d ProcessGroup APIs. See this issue: #39662. It might be helpful to sync with Alex offline to make sure there is no surprises and conflicts in our plans.

cc @agolynski is this consistent with the c10d API cleanup plan?

Sergei-Lebedev commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -31,4 +31,76 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
+  for (int r = 0; r < numRanks; r++) {
+    int sendRank = (myRank + r) % numRanks;
+    int recvRank = (myRank - r + numRanks) % numRanks;
+    C10D_NCCL_CHECK(ncclSend(
+        ((char*)sendbuff) + sendRank * rankOffset,
+        count,
+        type,
+        sendRank,
+        comm,
+        stream));
+    C10D_NCCL_CHECK(ncclRecv(
+        ((char*)recvbuff) + recvRank * rankOffset,
+        count,
+        type,
+        recvRank,
+        comm,
+        stream));
```
Hi @srinivas212, according to [nccl documentation](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/p2p.html#point-to-point-communication) perhaps these ncclSend and ncclRecv calls must be surrounded with ncclGroupStart() ncclGroupEnd() to avoid deadlock

srinivas212 replied:
Hi @Sergei-Lebedev - you are right. I'm relying on nccl_group_guard in ProcessGroupNCCL.cpp to ensure send/recv gets called within group semantics. 

https://github.com/srinivas212/pytorch/blob/8f78ccbe54af2b13a8b20a0016d7faf907a5253d/torch/lib/c10d/ProcessGroupNCCL.cpp#L740

For sake of clarity, I can add an explicit group start / group end. Nested group start/end does not affect performance AFAIK and after discussing with @sjeaugey we agreed this could be a good example for folks looking for alltoall examples as well.

Sergei-Lebedev replied:
I see, thanks for clarification. One more question, since GroupStart/GroupEnd calls are outside of loop it means that nccl scheduler will process bunch of send/recv so these tricky indexes for sendRank and recvRank are not necessary here

srinivas212 replied:
Yes, I have fixed the loop as well with the latest PR.

agolynski commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -31,4 +31,76 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
       getNcclVersion();
 }
 
+#if NCCL_VERSION_MIN(2, 7, 0)
```
nit: maybe NCCL_VERSION_GE?

agolynski commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -31,4 +31,76 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
+  C10D_NCCL_CHECK(ncclCommCount(comm, &numRanks));
+  C10D_NCCL_CHECK(ncclCommUserRank(comm, &myRank));
+  size_t rankOffset = count * elem_size;
+  for (int r = 0; r < numRanks; r++) {
```
nit: maybe call it rank_diff?

agolynski commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -911,6 +987,79 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::barrier(
+        });
+  }
+#else
+  throw std::runtime_error("ProcessGroupNCCL does not support alltoall_base");
```
nit: ProcessGroupNCCL only supports alltoall* for NCCL lib version >= 2.7.0

mrshenli commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -31,4 +31,62 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
       getNcclVersion();
 }
 
+#if NCCL_VERSION_GE(2, 7, 0)
```
@agolynski any reason these functions should be added to `NCCLUtils` instead of `ProcessGroupNCCL`? Do you plan to do this for all c10d APIs when working on #39662?

mrshenli commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -48,6 +54,25 @@ namespace c10d {
 
 std::string getNcclVersion();
 std::string ncclGetErrorWithVersion(ncclResult_t error);
+ncclResult_t gpuAlltoall(
```
as this is a NCCL file, we don't need the `gpu` prefix in the name?

mrshenli commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -17,6 +17,12 @@
 #endif
 
+#define NCCL_VERSION_GE(major, minor, patch) \
+  ((NCCL_MAJOR > major) ||                   \
```
Do we need to check `defined(NCCL_MAJOR)` and `defined(NCCL_MINOR)`?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -537,6 +537,16 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
 namespace {
 
+// Single tensor version of check_gpu_tensors()
+void check_gpu_tensor(const at::Tensor& tensor) {
```
should the `check_gpu_tensors` call this function to avoid code duplication?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -616,6 +626,72 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
   return flattened;
 }
 
+void checkSplitSizes(
```
This is a duplication of the `checkSplitSizes` in `ProcessGroupMPI.cpp`? If so, let's move it to some common place, e.g., a protected method in `ProcessGroup`, to avoid duplication..

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -616,6 +626,72 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
+    const at::Tensor& tensor,
+    int group_size) {
+  if (split_sizes.size() == 0) {
+    if (tensor.size(0) % group_size != 0) {
```
Let's use `TORCH_CHECK` instead

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -616,6 +626,72 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
+  }
+}
+
+int64_t computeLengthsAndOffsets(
```
This function and the function below are also duplication with the ones in `ProcessGroupMPI.cpp`?

mrshenli reviewed (commented):
I understand that we cannot add tests for now as PyTorch is still on 2.4.8. 

@agolynski shall we fix test error in #40622 and land so that tests can be added to this PR?

mrshenli commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -31,4 +31,62 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
+    void* sendbuff,
+    void* recvbuff,
+    size_t count,
+    size_t elem_size,
```
nit: all other part of this file seem to use camel instead of snake naming convention.

mrshenli commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -48,6 +54,25 @@ namespace c10d {
+ncclResult_t gpuAlltoallv(
+    void* sendbuff,
+    const int* sendcounts,
+    const int* sdispls,
```
curious, what does `displs` stand for?

srinivas212 replied:
Displacements: http://www.mpich.org/static/docs/latest/www3/MPI_Alltoallv.html

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -289,6 +301,29 @@ class ProcessGroupNCCL : public ProcessGroup {
   void parseNcclBlockingWait();
 
+   // NCCL alltoall implemented using NCCL send/recv
+   ncclResult_t ncclAlltoall(
```
do we also need to use ifdef to guard this?

srinivas212 replied:
Moved to anonymous namespace in C++ file.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -469,6 +469,64 @@ std::exception_ptr ProcessGroupNCCL::checkForNCCLErrorsInternal(
+  return ncclSuccess;
+}
+
+ncclResult_t ProcessGroupNCCL::ncclAlltoallv(
```
Looks like these two are only used in this file. I wonder if it makes sense to put them in an anonymous namespace in the cpp file and not mention them in the hpp file as private methods?

srinivas212 replied:
Fixed.

mrshenli reviewed (commented):
This PR LGTM. I only left two minor comments. Please sync with @agolynski on the plans for tests.

mrshenli commented:
Test failures are irrelevant to this PR. The culprit #40026 has been reverted on master.

mrshenli commented on test/distributed/test_distributed.py:
```diff
@@ -1540,49 +1562,147 @@ def _test_all_to_all_helper(self, group, group_id, rank):
         self._test_all_to_all_helper(group, group_id, rank)
 
-    @unittest.skipIf(BACKEND != "mpi", "Only MPI supports all_to_all_single")
+    @unittest.skipIf(BACKEND != "mpi", "Only MPI supports CPU tensors")
```
nit: CPU tensors -> CPU all_to_all_single?

mrshenli reviewed (approved):
Thanks for adding this feature!

Commit:
Add NCCL Alltoall to PT NCCL process group (#39984)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39984

Add Alltoall and Alltoallv to PT NCCL process group using NCCL Send/Recv.

Reviewed By: jiayisuse

Differential Revision: D20781624

fbshipit-source-id: 25cc94615be34d73954103c7e97b3ab97672c736

mrshenli commented:
This breaks master build. Sorry, I have to revert this. See:

https://app.circleci.com/pipelines/github/pytorch/pytorch/192460/workflows/a595ad03-dc27-47b7-9690-fbbef6a6489d/jobs/6320512

https://app.circleci.com/pipelines/github/pytorch/pytorch/192460/workflows/a595ad03-dc27-47b7-9690-fbbef6a6489d/jobs/6320506
